src/docs/arch-logging.dox

```diff
@@ -1,8 +1,167 @@
+logging subsystem. The main thread, called \c log_server, is a general thread. It
+flushes out the current slot to the file system. It does this every 50 milliseconds
+and this is done so that if a system becomes idle, buffered log records in the memory
+of a slot do not languish. The thread also performs log archiving to remove any log
+files that are no longer needed based on the checkpoint LSN. Finally the main thread
+performs log file pre-allocation, if configured. The pre-allocation algorithm is
+adaptable. If the thread did not make sufficient numbers of pre-allocated log files
+the system then makes more. If too few were used, then later pre-allocation will
+create fewer of them.
+
+There is another internal thread dedicated to managing operations on log files. This
+is the \c log_file_server thread. I/O operations can be slow and costly so this
+thread takes the burden of closing and syncing files to disk off of user application threads.
+If a log file needs to be closed, then the application thread moves its file handle to the
+\c log->log_close_fh field and signals this thread. This thread will fsync the file
+handle, and then close it. Updating the \c log_close_lsn and the \c sync_lsn values.
+
+The third internal thread manages advancing the \c write_lsn in proper LSN order.
+That LSN must be advanced in order. Therefore it becomes a point of single threading
+in the path of writing a log record. The goal is to avoid single threading application
+threads. So after a log slot is written to the file system, the application thread
+only needs to set the slot's state to a special state \c WT_LOG_SLOT_WRITTEN and
+return. This internal thread then walks the array of log slots looking for written
+slots and advances the \c write_lsn in the proper order. Along the way the thread
+will coalesce adjacent LSNs of slots in order to advance the LSN in larger increments.
+This thread will then clean up and mark the state of the processed slots
+as \c WT_LOG_SLOT_FREE.
+
+@subsection Using the log
+
+In addition to storing a record of the operations performed to data in the write-ahead
+log, users can write their own strings into the logs via the \c WT_SESSION::log_printf
+method. That method takes a printf-style format string and arguments and will write
+that string into the write-ahead log as a message record type.
+
+Transactions are stored in the write-ahead log and their purpose is to allow recovery
+in the case of crash or failure. When the system is restarted, WiredTiger will always
+run recovery. If the system previously shutdown cleanly, then recovery determines that
+the log does not need to be replayed. Otherwise, recovery replays all operations from
+the LSN of the most recent checkpoint. When timestamps are in use, recovery also
+performs a rollback-to-stable operation on tables that are not logged.
+# Remove this line and uncomment the next line after the rollback to stable page exists.
```

Comment:
[This page exists](https://github.com/wiredtiger/wiredtiger/blob/develop/src/docs/arch-rts.dox).